improver_tests/threshold/test_ThresholdPytest.py

```diff
@@ -0,0 +1,599 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the threshold.Threshold plugin."""
+
+
+import numpy as np
+import pytest
+from iris.coords import CellMethod, DimCoord
+from iris.cube import Cube
+
+from improver.threshold import Threshold as Threshold
+
+
+@pytest.mark.parametrize(
+    "kwargs,exception",
+    [
+        # a threshold of zero is used with a multiplicative fuzzy factor
+        (
+            {"threshold_values": 0.0, "fuzzy_factor": 0.6},
+            "Invalid threshold with fuzzy factor",
+        ),
+        # a fuzzy factor of minus 1 is given
+        (
+            {"threshold_values": 0.6, "fuzzy_factor": -1.0},
+            "Invalid fuzzy_factor: must be >0 and <1: -1.0",
+        ),
+        # a fuzzy factor of zero is given
+        (
+            {"threshold_values": 0.6, "fuzzy_factor": 0.0},
+            "Invalid fuzzy_factor: must be >0 and <1: 0.0",
+        ),
+        # a fuzzy factor of unity is given
+        (
+            {"threshold_values": 0.6, "fuzzy_factor": 1.0},
+            "Invalid fuzzy_factor: must be >0 and <1: 1.0",
+        ),
+        # a fuzzy factor of 2 is given
+        (
+            {"threshold_values": 0.6, "fuzzy_factor": 2.0},
+            "Invalid fuzzy_factor: must be >0 and <1: 2.0",
+        ),
+        # fuzzy_factor and fuzzy_bounds both set
+        (
+            {"threshold_config": {"0.6": [0.4, 0.8]}, "fuzzy_factor": 2.0},
+            "Invalid combination of keywords",
+        ),
+        # fuzzy_bounds contains one value
+        (
+            {"threshold_config": {"0.6": [0.4]}},
+            "Invalid bounds for one threshold: \\(0.4,\\).",
```

Comment:
[Optional] You can reduce the double-escapes to single-escapes with regex-strings. It isn't a massive improvement though.
```suggestion
            r"Invalid bounds for one threshold: \(0.4,\).",
```

Reply:
New-fangled (by which I mean unfamiliar). I'll leave this as is so I understand what I've done.